[clang-doc] Refactor error handling to use ExitOnError

snarang181 opened the pull request:
Fixes #140085
This patch refactors the clang-doc tool to replace manual error handling (e.g., checking error codes and calling exit()) with llvm::ExitOnError. 

Addresses:
- https://github.com/llvm/llvm-project/issues/140085

snarang181 committed:
[clang-doc] Refactor error handling to use ExitOnError

This patch replaces manual error checks and exit() calls in clang-doc
with llvm::ExitOnError for consistency and maintainability.

No functional changes to outputs or APIs.

ilovepi commented on clang-tools-extra/clang-doc/tool/ClangDocMain.cpp:
```diff
@@ -49,6 +49,7 @@
 using namespace clang::ast_matchers;
 using namespace clang::tooling;
 using namespace clang;
+static llvm::ExitOnError ExitOnErr;
```
```suggestion

static llvm::ExitOnError ExitOnErr;
```
Nit: newline between using and the declaration

ilovepi replied:
actually, can you move this to be below all the cl::opts and constants? around line 131 seems like an OK spot, but I'll leave the precise location to you.

snarang181 replied:
Moved before the function definitions.

ilovepi commented on clang-tools-extra/clang-doc/tool/ClangDocMain.cpp:
```diff
@@ -371,22 +367,14 @@ Example usage for a project using a compile commands database:
   // Ensure the root output directory exists.
-  if (std::error_code Err = llvm::sys::fs::create_directories(OutDirectory);
-      Err != std::error_code()) {
-    llvm::errs() << "Failed to create directory '" << OutDirectory << "'\n";
-    return 1;
-  }
```
Does errorCodeToError give diagnostics for the directory failure? I don't recall that it does, and I'd like to keep the diagnostic message if possible.

snarang181 replied:
I have tried to maintain the same semantics by using `ExitOnErr`.  After this change, for the lit test I added, which tries to create a new directory under `root`, the clang-doc error I get is `clang-doc error: Failed to create directory '/root/docs': Read-only file system`.

ilovepi commented:
Other than some nits the patch is mostly ready to go. It does need a test though, which sadly we don't already have. Please add a lit test under clang-tools-extra/tests/clang-doc.

ilovepi commented:
oh, can update the description? Not all of that belongs in the text of the commit body.

1. Use `Fixes #140085` which will allow the issue to be auto closed once this lands
2.  drop the bit at the bottom from "No functional ..." since that doesn't need to be in the commit log
3. instead of bullet points, you can just summarize those points in a short paragraph, or omit altogether, since its linked to the bug which outlines all of that.

ilovepi commented:
Also, thanks for the patch! 😄

snarang181 committed:
Addressing PR feedback

snarang181 committed:
Clang-format fixes

ilovepi commented on clang-tools-extra/test/clang-doc/invalid-file-error.test:
```diff
@@ -0,0 +1,4 @@
+// RUN: not clang-doc %S/Inputs/basic-project/src/Circle.cpp -output=/root/docs 2>&1 | FileCheck %s
```
You can't use /root as the output path. Use %t, which is will be substituted w/ a temporary file path unique to this test. since that dir won't exist, the test should fail. If that doesn't quite fit, you can use the `rm -rf %t && mkdir %t` in other tests, then you can use `%t/some/invalid/path` for the output.

You don't need to use a real file for this test though. Rename this file `invalid-options.cpp`, and then just only have your RUN and CHECK lines, then you can drop the reference to Circle.cpp, and just use `%s`. 

I also think you need more tests. Ideally, one test, both positive and negative for each of the error cases you've updated. Probably all of those will need their own RUN lines, and the CHECK lines will probably have to be replaces with their own prefix. Several of the other tests demonstrate the pattern you can follow with
`--check-prefix=INVALID-FILE`, and then instead of `CHECK:` you can write `INVALID-FILE:`. They can all go in this file, which is why I'd suggest naming it something related to what is being tested. `driver-error.cpp` or something is also a reasonable choice.

snarang181 replied:
Sorry, I might be misunderstanding but won't `llvm::sys::fs::create_directories()` create a valid directory path? 
Giving the `/root/` output directory fails because it is a read-only file system. If I take the `%t` approach, the clang-doc error will not be fired.

ilovepi replied:
Well, I had forgotten that `create_directories()` works differently than `create_directory()`,  in that its basically `mkdir -p`. 

However, in general, you cannot rely on `/root` failing. Imagine if this ran in a docker container which often runs as the root user. That operation would succeed.  Many of our bots in CI run this way.  Using `/dev/null` may work, since I think we may do some kind of remapping for that in the internal shell used on Windows. However, I don't recall precisely, and I seldom build on Windows. Failing that, we'd need to use %t and change permissions on it. I'd rather avoid trying to modify permission in a test  though.

Looking through the tree, https://github.com/llvm/llvm-project/blob/main/llvm/test/tools/llvm-ifs/fail-file-write.test and the windows test next to it may be the best examples to follow, if we have to use permission modifications in the test.

ilovepi commented on clang-tools-extra/test/clang-doc/invalid-file-error.test:
```diff
@@ -0,0 +1,4 @@
+// RUN: not clang-doc %S/Inputs/basic-project/src/Circle.cpp -output=/root/docs 2>&1 | FileCheck %s
+
+// CHECK: clang-doc error: Failed to create directory
```
CI says this check is not passing, so do check why. I'm guessing its the file path to Circle.cpp, so it may start working better if you switch to %s.

ilovepi commented on clang-tools-extra/clang-doc/tool/ClangDocMain.cpp:
```diff
@@ -373,20 +370,16 @@ Example usage for a project using a compile commands database:
-    llvm::errs() << "Failed to create directory '" << OutDirectory << "'\n";
-    return 1;
+    ExitOnErr(llvm::createStringError(
+        llvm::inconvertibleErrorCode(), "Failed to create directory '%s': %s",
```
Is inconvertibleErrorCode, the right fit. I guess do you want/need a string error at all? IIRC there are some error APIs in Error.h that are about file system errors.  `createFileError(...)` I think

snarang181 replied:
The string error looks verbose and explanatory to me. 
However, would you prefer me looking into the error APIs or revert this change altogether?

ilovepi replied:
So the first point stands no matter what, since it's certainly not an incovertibleErrorCode. 

I'm wondering if we wouldn't be better served by using the createFileError API I mentioned, which I believe will give a good error message. If the other API isn't sufficient, then we can keep the string error and choose a better error code type than the inconvertible error code. 

My preference is for you to check the createFileError() API, and determine if its a good enough fit. We're using it elsewhere IIRC, so I assume its giving a good enough diagnostic. That said, I don't recall offhand what the output would be, so we should check.

snarang181 committed:
Use more relevant file system API for error handling
Address issue with the lit tests

snarang181 committed:
Formatting commit

snarang181 committed:
Rename test file to reflect multiple test cases

snarang181 committed:
Modifying expected result to iron out cross-platform output diff

ilovepi commented on clang-tools-extra/clang-doc/tool/ClangDocMain.cpp:
```diff
@@ -284,10 +284,7 @@ Example usage for a project using a compile commands database:
-      return 1;
-    }
+    ExitOnErr(getHtmlAssetFiles(argv[0], CDCtx));
   }
```
I think you may need to rebase your patch, since we've landed experimental `mustache` support. There should be a similar `if (format == "mustache")` section if you look at the current main.

snarang181 replied:
I am not familiar with this, will have to look into it.

ilovepi commented on clang-tools-extra/clang-doc/tool/ClangDocMain.cpp:
```diff
@@ -371,22 +368,15 @@ Example usage for a project using a compile commands database:
+  ExitOnErr(G->generateDocs(OutDirectory, std::move(USRToInfo), CDCtx));
   llvm::outs() << "Generating assets for docs...\n";
-  Err = G->get()->createResources(CDCtx);
+  Err = G->createResources(CDCtx);
```
this seems unrelated. That said, should we just use `ExitOnError` here too? that seems better than printing a warning and then exiting. I don't recall why we don't return failure if an error happens creating resources ... for most backends that's a rather big deal. 

If all the tests pass with this using `ExitOnError`, then lets do that, otherwise lets drop this bit, and add a TODO for us to figure out how to handle this case more gracefully.

snarang181 replied:
I will try this out.

ilovepi reviewed (commented):
Thanks for the work on this. I think this is getting pretty close, but I think there are more test cases required.

Tests that we need:
1. Generator name
2. HTML asset file that's missing
3. A mapping failure (there may already be tests that handle this)
4. Failing to create directories
5. Failure to generateDocs
6. Failure from createResources

If there are things that are more difficult to test, then we should have some TODOs in the code and file a tracking bug in the issue tracker. I think the generateDocs failures could be in the hard to test scenario w/o adding something like a test generator/backend. Something like that is out of scope for this patch, and should be handled as a follow up, so I doubt we'll get all of them in this patch.  But the asset directory should be pretty easy to add another check for.

If you hadn't seen it the programmers manual has some advice for ExitOnError https://llvm.org/docs/ProgrammersManual.html#using-exitonerror-to-simplify-tool-code, and a lot to say about error handling in general.

snarang181 commented:
Also, from the CI test run, I see the Linux build failing. This seems unrelated, right? Does it warrant a rerun? 
```
/home/gha/actions-runner/_work/llvm-project/llvm-project/libc/test/src/math/smoke/HypotTest.h:31: FAILURE
Failed to match aNaN against LIBC_NAMESPACE::testing::getMatcher< LIBC_NAMESPACE::testing::TestCond::EQ>(func(sNaN, neg_inf)).
Expected floating point value: (+Infinity)
Actual floating point value: (NaN)
[  FAILED  ] LlvmLibcHypotf16Test.SpecialNumbers
Ran 1 tests.  PASS: 0  FAIL: 1/home/gha/actions-runner/_work/llvm-project/llvm-project/libc/test/src/math/smoke/HypotTest.h:31: FAILURE
Failed to match aNaN against LIBC_NAMESPACE::testing::getMatcher< LIBC_NAMESPACE::testing::TestCond::EQ>(func(sNaN, neg_inf)).
Expected floating point value: (+Infinity)
Actual floating point value: (NaN)
[  FAILED  ] LlvmLibcHypotf16Test.SpecialNumbers
Ran 1 tests.  PASS: 0  FAIL: 1
```

snarang181 commented on clang-tools-extra/test/clang-doc/invalid-options.cpp:
```diff
@@ -0,0 +1,12 @@
+
+//
+// Test: Invalid format option
+// RUN: not clang-doc %s -format=badformat 2>&1 | FileCheck --check-prefix=BAD-FORMAT %s
```
Thanks for letting me know.

ilovepi commented:
> @ilovepi, thanks for the patience and taking out the time to give detailed reviews. I will work on addresing your comments. One word on the tests: I was trying to test some of the other fail conditions but seems like a lot of the error conditions are caught by the parsing early on and do not actually invoke our error handler (like the current existing test case). Would adding those cases be OK regardless? Even though this patch does not directly address those issues. In that vein, it might be better to add `clang-doc` tests as part of a different PR, and I will be happy to work on that follow up too. Let me know what you think.

Its no problem. If we don't mentor folks to be better contributors, we won't get new ones, and if people didn't take the time to explain the best practices to *me* in my own reviews, I wouldn't be able to do this either.

As for tests if we're exercising those cases, then its fine if another mechanism triggers the error in something like the options parsing code from cl::opt. For this patch, I'd like to see a test for each of the places we added `ExitOnErr()` calls (at least to the extent that's practical). If its not trivial to do so, lets add TODOs and file Issues on GitHub as necessary, so that work doesn't get lost/forgotten. If you have a follow up patch, then you can remove the TODOs as you address them.

snarang181 committed:
Modifications in ClangDocMain.cpp to handle new rebased code
Some more cleanup -- redundant code removed
Cleanup of test cases according to conventions

snarang181 committed:
Add more test cases for clang-doc failures

snarang181 commented:
> Tests that we need:
> 
> 1. Generator name
> 2. HTML asset file that's missing
> 3. A mapping failure (there may already be tests that handle this)
> 4. Failing to create directories
> 5. Failure to generateDocs
> 6. Failure from createResources
> 

@ilovepi, 
For Case 1, I cannot simulate an invalid generator name test case. For the invalid generator name case to be triggered, we need to first pass the valid format option, and then the generator name is checked. If we pass a bad format, we fail early (there is a test case covering this). Currently, all formats have registered generators; so maybe the bad format option is an OK test case for now? 

For Case 2, HTML needs a .js file. If it does not find any, it tries to use `index.js` in the `assets/` dir, and just warns out that it is using a fallback file. I have added a test case for this but with a warning only as the command successfully returns. 

For Case 3, added two tests: one with `ignore-map-errors=false` which errors out with the `ExitOnErr` handler and the other with the default `true` value of `ignore-map-errors` which gives a map warning and continues. 

For Case 4, we had a test. 

For Cases 5 and 6, I am not sure on how we can test these right off the bat. All ideas I can think of involve patching the code with a hook to make a certain step fail so that we can reach that error handling condition, which of course is not possible from a CI perspective. I added a `TODO` in the test file. 

Also, addressed other changes. Let me know how it looks.

ilovepi commented:
> @ilovepi, For Case 1, I cannot simulate an invalid generator name test case. For the invalid generator name case to be triggered, we need to first pass the valid format option, and then the generator name is checked. If we pass a bad format, we fail early (there is a test case covering this).

If we have coverage of that, then its fine if it doesn't trigger the precise code path. That does kind of make me wonder if we need that code at all though... there's `cantFail` ... but I'm hesitant to add that anywhere.

> 
> For Case 2, HTML needs a .js file. If it does not find any, it tries to use `index.js` in the `assets/` dir, and just warns out that it is using a fallback file. I have added a test case for this but with a warning only as the command successfully returns.

Hmm, that's OK if we can't trigger the behavior, but again if we can't hit the exit case, maybe we don't need the code. For now lets leave a TODO and move on.

> 
> For Case 3, added two tests: one with `ignore-map-errors=false` which errors out with the `ExitOnErr` handler and the other with the default `true` value of `ignore-map-errors` which gives a map warning and continues.
> 

Perfect :)

> For Case 4, we had a test.
> 
> For Cases 5 and 6, I am not sure on how we can test these right off the bat. All ideas I can think of involve patching the code with a hook to make a certain step fail so that we can reach that error handling condition, which of course is not possible from a CI perspective. I added a `TODO` in the test file.
> 

That's fine. There's probably some way we can structure the input to trigger those cases, but it may be hard/complex, and this patch is still an improvement over the status quo.

> Also, addressed other changes. Let me know how it looks.

snarang181 committed:
Add explicit output dir in test case

snarang181 committed:
Wrap executor in ExitOnErr and modify uses
Add test case

snarang181 committed:
Run clang-format

snarang181 commented:
@ilovepi, wrapped the creation of the `Executor` in the error handler, created a helper like you suggested, and added a test case for the creation of the `Executor` failing.

ilovepi commented on clang-tools-extra/clang-doc/tool/ClangDocMain.cpp:
```diff
@@ -247,6 +247,22 @@ sortUsrToInfo(llvm::StringMap<std::unique_ptr<doc::Info>> &USRToInfo) {
 }
 
+llvm::Error runMappingPhase(tooling::ToolExecutor *Executor,
+  clang::doc::ClangDocContext &CDCtx, tooling::ArgumentsAdjuster &ArgAdjuster, bool IgnoreMappingFailures ) {
```
I was thinking more of an API that took in the Error from the executor and returned something usable with ExitOnErr.

```cpp
ExitOnErr(helper_function(Executor->execute(doc::newMapperActionFactory(CDCtx), ArgAdjuster));
```

ilovepi commented on clang-tools-extra/test/clang-doc/invalid-options.cpp:
```diff
@@ -0,0 +1,29 @@
+// CHECK: {{(Not a directory|no such file or directory)}}
+
+/// Invalid format option.
+// RUN: not clang-doc %s --output=%t.dir -format=badformat 2>&1 | FileCheck %s --check-prefix=BAD-FORMAT
```
If you're going to use %t.dir, you have to `rm -rf %t.dir` at the top of the test. Alternativly, you can rewrite the preamble to make things easier

```
RUN: rm -rf %t && mkdir %t && touch %t/file
```
Then in the checks you can use %t/file if you need a dir that isn't a dir.

ilovepi commented on clang-tools-extra/test/clang-doc/invalid-options.cpp:
```diff
@@ -0,0 +1,29 @@
+// RUN: clang-doc %t/nonexistent.cpp 2>&1 | FileCheck %s --check-prefix=MAP-WARN
+// MAP-WARN: Error mapping decls in files. Clang-doc will ignore these files and continue
+
+///Invalid executor type
+// RUN: not clang-doc --executor=invalid %s 2>&1 | FileCheck %s --check-prefix=EXECUTOR-FAIL
+// EXECUTOR-FAIL: clang-doc error:
+// EXECUTOR-FAIL: Executor "invalid" is not registered
```
There are other ways to make the executor fail besides passing a bad executor flag. I'm pretty sure passing extra args is one way. Can you try this test with some nonese compiler flags passed through --extra_args (or whatever name we use...)? That may give us a nicer signal, and trigger the ExitOnError code in the tool.

ilovepi commented on clang-tools-extra/test/clang-doc/invalid-options.cpp:
```diff
@@ -0,0 +1,29 @@
+/// Missing HTML asset directory (warning only).
+// RUN: clang-doc %s -format=html -asset=%t/nonexistent-assets 2>&1 | FileCheck %s --check-prefix=ASSET-WARN
+// ASSET-WARN: Asset path supply is not a directory
+// ASSET-WARN: falling back to default
```
We can drop this line. Matching the first part is enough.

ilovepi commented on clang-tools-extra/clang-doc/tool/ClangDocMain.cpp:
```diff
@@ -245,10 +247,29 @@ sortUsrToInfo(llvm::StringMap<std::unique_ptr<doc::Info>> &USRToInfo) {
   }
 }
 
+llvm::Error runMappingPhase(tooling::ToolExecutor *Executor,
+                            clang::doc::ClangDocContext &CDCtx,
+                            tooling::ArgumentsAdjuster &ArgAdjuster,
+                            bool IgnoreMappingFailures) {
+  auto Err = Executor->execute(doc::newMapperActionFactory(CDCtx), ArgAdjuster);
+  if (Err) {
+    if (IgnoreMappingFailures) {
+      llvm::errs() << "Error mapping decls in files. Clang-doc will ignore "
+                      "these files and continue:\n"
+                   << toString(std::move(Err)) << "\n";
+      return llvm::Error::success();
+    }
+    return Err;
+  }
+  return llvm::Error::success();
+}
```
When I described this before, I thought I had mentioned using a helper for the error handling.

So something along these lines that still makes what's happening clear, but hides the details about IngoreMappingBehavior.

```cpp
ExitOnErr(helper(Executor->execute(doc::newMapperActionFactory(CDCtx), ArgAdjuster));
```

snarang181 committed:
Add helper for handling mapping phase errors

snarang181 committed:
Incorporating suggestions in the lit test

snarang181 committed:
Add helper function for handling error handling
in case of output dir existence too

snarang181 committed:
Remove same comment

ilovepi reviewed (approved):
I think this will be fine to land after the small changes I've requested are done, and CI is passing. Thanks for helping out clang-doc, and bearing with the review process.

snarang181 commented:
Thanks again @ilovepi for the handholding, appreciate it. 

BTW -- do you know what's going on with the Linux CI?

snarang181 committed:
Renaming functions, making them static, some more cleanup
